blocks/serialization.py

```diff
@@ -1,65 +1,181 @@
+        ``True``.
+    allow_duplicates : bool, optional
+        Allow multiple shared variables to have the same name, in which
+        case they will be numbered e.g. `x`, `x_2`, `x_3`, etc. Defaults to
+        ``True``.
+
+    Raises
+    ------
+    ValueError
+        If an unnamed shared variable is encountered and `allow_unnamed` is
+        ``False``, or if two shared variables have the same name, and
+        `allow_duplicates` is ``False``.
 
-Python is unable to serialize nested functions. Make sure that all your \
-functions are available in the global namespace."""
+    """
+    def __init__(self, zip_file, allow_unnamed=True, allow_duplicates=True):
+        super(PersistentParameterID, self).__init__(zip_file)
+        self.name_counter = defaultdict(int)
+        self.ndarray_names = {}
+        self.allow_unnamed = allow_unnamed
+        self.allow_duplicates = allow_duplicates
+
+    def _resolve_name(self, obj):
+        if id(obj) in self.ndarray_names:
+            name = self.ndarray_names[id(obj)]
+            count = self.name_counter[name]
+            if count:
+                if not self.allow_duplicates:
+                    raise ValueError("multiple shared variables with the name "
+                                     "`{0}` found".format(name))
+                name = '{0}_{1}'.format(name, count + 1)
+            self.name_counter[name] += 1
+            return name
+        return super(PersistentParameterID, self)._resolve_name(obj)
+
+    def __call__(self, obj):
+        if isinstance(obj, SharedVariable):
+            if obj.name:
+                if obj.name == 'pkl':
+                    raise ValueError(
+                        "can't pickle shared variable with name `pkl`")
+                if hasattr(obj.tag, 'annotations'):
+                    name = BRICK_DELIMITER.join(
+                        [brick.name for brick
+                         in obj.tag.annotations[0].get_unique_path()] +
+                        [obj.name])
+                else:
+                    name = obj.name
+                self.ndarray_names[id(obj.container.storage[0])] = name
+            elif not self.allow_unnamed:
+                raise ValueError("unnamed shared variable, {}".format(obj))
+        return super(PersistentParameterID, self).__call__(obj)
+
+
+class PicklerWithWarning(pickle.Pickler):
+    if six.PY2:
+        dispatch = pickle.Pickler.dispatch.copy()
+    else:
+        dispatch_table = copyreg.dispatch_table.copy()
+
+    def save_global(self, obj, **kwargs):
+        module = getattr(obj, "__module__", None)
+        if module == '__main__':
+            warnings.warn(
+                MAIN_MODULE_WARNING.format(kwargs.get('name', obj.__name__))
+            )
+        pickle.Pickler.save_global(self, obj, **kwargs)
+    if six.PY3:
+        dispatch_table[six.types.FunctionType] = save_global
+    else:
+        dispatch[six.types.ClassType] = save_global
```

Comment:
We might be better off fixing all those empty dictionaries in the log if it
turns out to be slow, because I do think that this serialisation method is
a big improvement over what we have right now. Also, I don't think that
serialisation is too time critical (if it is, it means you're checkpointing
too often probably).
On May 18, 2015 4:21 PM, "Dmitry Bogdanov" <notifications@github.com> wrote:

> In blocks/serialization.py
> <https://github.com/mila-udem/blocks/pull/615#discussion_r30543581>:
>
> > +    if six.PY2:
> > +        dispatch = pickle.Pickler.dispatch.copy()
> > +    else:
> > +        dispatch_table = copyreg.dispatch_table.copy()
> > +
> > +    def save_global(self, obj, **kwargs):
> > +        module = getattr(obj, "__module__", None)
> > +        if module == '__main__':
> > +            warnings.warn(
> > +                MAIN_MODULE_WARNING.format(kwargs.get('name', obj.__name__))
> > +            )
> > +        pickle.Pickler.save_global(self, obj, **kwargs)
> > +    if six.PY3:
> > +        dispatch_table[six.types.FunctionType] = save_global
> > +    else:
> > +        dispatch[six.types.ClassType] = save_global
>
> I will check the exact slowdown tomorrow. I am worried that the training
> log containing lots of small objects will take forever to be handled by
> pickle.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mila-udem/blocks/pull/615/files#r30543581>.
>